src/flash.rs

```diff
@@ -34,3 +49,436 @@ impl ACR {
+        // 4. Check that the new embedded Flash memory clock source and/or the new AXI bus clock prescaler value are taken in account by reading back the embedded Flash memory clock source status and/or the AXI bus prescaler value in the RCC_CFGR register of the reset and clock controller (RCC).
+        // To decrease the embedded Flash memory clock source frequency:
+        // 1. Modify the embedded Flash memory clock source and/or the AXI bus clock prescaler in the RCC_CFGR register of reset and clock controller (RCC).
+        // 2. Check that the embedded Flash memory new clock source and/or the new AXI bus clock prescaler value are taken into account by reading back the embedded Flash
+        // memory clock source status and/or the AXI interface prescaler value in the RCC_CFGR register of reset and clock controller (RCC).
+        // 3. If necessary, program the LATENCY and WRHIGHFREQ bits to the right value in FLASH_ACR register, as described in Table 17.
+        // 4. Check that the new number of wait states has been taken into account by reading back the FLASH_ACR register.
+    }
+
+    /// Unlock the flash memory, must only be executed on locked flash
+    fn unlock(&self, bank: Bank) -> Result<(), Error> {
+        const FLASH_KEY1: u32 = 0x4567_0123;
+        const FLASH_KEY2: u32 = 0xCDEF_89AB;
+
+        let regs = self.bank_registers(bank);
+        regs.keyr.write(|w| unsafe { w.bits(FLASH_KEY1) });
+        regs.keyr.write(|w| unsafe { w.bits(FLASH_KEY2) });
+
+        if regs.cr.read().lock().bit_is_clear() {
+            Ok(())
+        } else {
+            Err(Error::Unlock)
+        }
+    }
+
+    /// Lock the flash memory, must be executed after an unlock
+    fn lock(&mut self, bank: Bank) {
+        let regs = self.bank_registers(bank);
+        regs.cr.modify(|_, w| w.lock().set_bit());
+    }
+
+    /// Read flash memory from a given bank, sector and offset by the length of buff / 4
+    ///
+    /// # Panics
+    ///
+    /// Panics if the sector is beyond the number of sectors supported by the user bank for this device
+    /// Panics if the buff length is not a multiple of a 32bit word
+    /// Panics if the start offset is not a multiple of a 32bit word
+    /// Panics if the buff length exceeds the user bank 2 end address for this device
+    pub fn read_sector(
+        &mut self,
+        bank: Bank,
+        sector: usize,
+        start: usize,
+        buff: &mut [u8],
+    ) -> Result<(), Error> {
+        if sector > bank.num_sectors() {
+            panic!("Sector out of bounds")
+        }
+        if buff.len() % 4 != 0 || start % 4 != 0 {
+            panic!("buff length and start offset must be a multiple of a 32bit word");
+        }
+
+        let regs = self.bank_registers(bank);
+
+        // Check that no Flash main memory operation is ongoing by checking the BSY bit
+        while regs.sr.read().bsy().bit_is_set() {}
```

Comment:
The RM description of the BSY bit refers to "write, erase or option byte" operations, not any operation. Specifically there can be (and likely are) flash read operations ongoing when BSY is clear. Could you update the comment? (also appears in write_sector and erase_sector)